File: src/games/pantomime/logic.ts
// src/games/pantomime/logic.ts — PURE game logic for Pantomime. No clock/RNG/IO; now/seed arrive in payloads.
import type { ColorToken, GameConfig, GameStateBase, PlayerId } from '../../sdk/types';
import type { DeckState } from '../../engine/deck';
import type { ScoreState } from '../../engine/scoring';
import type { TurnOrderState } from '../../engine/turnOrder';
import type { RevealGateState } from '../../engine/revealGate';
import type { TimerState } from '../../engine/timer';
import * as deck from '../../engine/deck';
import * as scoring from '../../engine/scoring';
import * as turnOrder from '../../engine/turnOrder';
import * as revealGate from '../../engine/revealGate';
import * as timer from '../../engine/timer';
import * as phaseMachine from '../../engine/phaseMachine';
import * as results from '../../engine/results';
import { asPlayerId } from '../../engine/ids';
import { buildPool } from './deck';
import { readOptions } from './config';
import type { PantomimeOptions } from './config';

export type PantomimePhase =
  | 'handoff' // "Pass the phone to <Actor> of <Team>"
  | 'reveal' // gate closed: actor taps to privately read the prompt
  | 'acting' // prompt revealed, timer running, team guessing
  | 'turnEnd' // turn over: show turn summary
  | 'results' // game over
  | 'error'; // bad config / empty pool (never reachable from a valid Setup)

export type TurnEndReason = 'timeExpired' | 'deckExhausted' | 'manualEnd';

export interface TurnEvent {
  promptId: string;
  result: 'correct' | 'skip';
}

export interface PantomimeTeamState {
  teamId: string;
  name: string;
  color?: ColorToken;
  playerIds: string[];
  /** Index into playerIds of who acts on this team's NEXT turn. */
  actorCursor: number;
  correctCount: number;
  skipCount: number;
}

export interface PantomimeTurnRecord {
  turnIndex: number; // global, 0-based
  roundIndex: number; // 0-based
  teamId: string;
  actorId: string;
  correct: number;
  skipped: number;
  endReason: TurnEndReason;
  promptIds: string[]; // resolved this turn (correct + skipped)
}

export interface PantomimeState extends GameStateBase {
  phase: PantomimePhase;
  options: PantomimeOptions;
  playerNames: Record<string, string>;
  /** Teams in fixed turn order. */
  teams: PantomimeTeamState[];
  /** Turn order over teamIds (index = active team, round = completed cycles). */
  turn: TurnOrderState;
  deck: DeckState<string>;
  currentPromptId: string | null;
  gate: RevealGateState;
  clock: TimerState;
  turnCorrect: number;
  turnSkipped: number;
  turnEvents: TurnEvent[];
  lastTurnEndReason: TurnEndReason | null;
  history: PantomimeTurnRecord[];
  score: ScoreState; // keyed by teamId
  /** Set when a target-score end condition fires; the round still finishes (fairness). */
  endRequested: boolean;
  winnerTeamIds: string[];
  errorCode: 'EMPTY_DECK' | 'BAD_TEAMS' | null;
}

export type PantomimeAction =
  | { type: 'HANDOFF_READY' }
  | { type: 'REVEAL' }
  | { type: 'START_ACTING'; now: number }
  | { type: 'TICK'; now: number }
  | { type: 'CORRECT' }
  | { type: 'SKIP' }
  | { type: 'PAUSE'; now: number }
  | { type: 'RESUME'; now: number }
  | { type: 'END_TURN_EARLY'; now: number }
  | { type: 'NEXT_TURN'; seed: number }
  | { type: 'RESET' };

const MACHINE = phaseMachine.defineMachine<PantomimePhase>({
  initial: 'handoff',
  nodes: {
    handoff: { id: 'handoff', to: ['reveal', 'turnEnd'] },
    reveal: { id: 'reveal', to: ['acting', 'turnEnd'] },
    acting: { id: 'acting', to: ['turnEnd'] },
    turnEnd: { id: 'turnEnd', to: ['handoff', 'results'] },
    results: { id: 'results', to: [], terminal: true },
    error: { id: 'error', to: [], terminal: true },
  },
});

const TEAM_COLORS: ColorToken[] = ['rose', 'sky', 'lime', 'gold'];

/** Seed used for in-game deck reshuffles. Deterministic (initial shuffle uses the host seed). */
const DRAW_SEED = 0;

function drawNext(d: DeckState<string>): { deck: DeckState<string>; promptId: string | null } {
  const r = deck.draw(d, 1, DRAW_SEED);
  return { deck: r.deck, promptId: r.drawn[0] ?? null };
}

function updateTeam(
  teams: PantomimeTeamState[],
  index: number,
  patch: (t: PantomimeTeamState) => PantomimeTeamState,
): PantomimeTeamState[] {
  return teams.map((t, i) => (i === index ? patch(t) : t));
}

export const activeTeam = (s: PantomimeState): PantomimeTeamState => s.teams[s.turn.index];
export const actorId = (s: PantomimeState): string => {
  const team = s.teams[s.turn.index];
  return team ? team.playerIds[team.actorCursor] : '';
};
export const actorName = (s: PantomimeState): string => s.playerNames[actorId(s)] ?? '';

function finalizeTurn(s: PantomimeState, reason: TurnEndReason, now: number): PantomimeState {
  const team = s.teams[s.turn.index];
  // Return any unresolved current prompt to the draw pile (not counted, not lost).
  let d = s.deck;
  if (s.currentPromptId !== null) {
    d = { ...s.deck, drawPile: [...s.deck.drawPile, s.currentPromptId] };
  }
  const record: PantomimeTurnRecord = {
    turnIndex: s.history.length,
    roundIndex: s.turn.round,
    teamId: team.teamId,
    actorId: team.playerIds[team.actorCursor],
    correct: s.turnCorrect,
    skipped: s.turnSkipped,
    endReason: reason,
    promptIds: s.turnEvents.map((e) => e.promptId),
  };
  return {
    ...s,
    phase: 'turnEnd',
    deck: d,
    currentPromptId: null,
    lastTurnEndReason: reason,
    history: [...s.history, record],
    clock: timer.pause(s.clock, now),
  };
}

export function createInitialState(config: GameConfig, seed: number): PantomimeState {
  const options = readOptions(config);
  const players = config.players;
  const playerNames: Record<string, string> = {};
  players.forEach((p) => {
    playerNames[p.id] = p.name;
  });

  const teamList = config.teams?.teams ?? [];
  const teams: PantomimeTeamState[] = teamList.map((t, i) => ({
    teamId: t.id as string,
    name:
      typeof t.name === 'string'
        ? t.name
        : t.name
          ? (t.name[config.lang] ?? t.name.en)
          : `Team ${i + 1}`,
    color: TEAM_COLORS[i % TEAM_COLORS.length],
    playerIds: t.memberIds.map((id) => id as string),
    actorCursor: 0,
    correctCount: 0,
    skipCount: 0,
  }));

  const pool = buildPool(options);
  const deckState = deck.create(
    pool.map((p) => p.id),
    seed,
  );
  const teamIds = teams.map((t) => t.teamId);
  const turn = turnOrder.init(teamIds as unknown as PlayerId[], 'circular', seed);

  const badTeams = teams.length < 2 || teams.some((t) => t.playerIds.length < 2);
  const empty = pool.length === 0;
  const errorCode: PantomimeState['errorCode'] = badTeams
    ? 'BAD_TEAMS'
    : empty
      ? 'EMPTY_DECK'
      : null;

  return {
    v: 1,
    phase: errorCode ? 'error' : 'handoff',
    finished: false,
    options,
    playerNames,
    teams,
    turn,
    deck: deckState,
    currentPromptId: null,
    gate: revealGate.init([]),
    clock: timer.create('countdown', options.roundSeconds * 1000),
    turnCorrect: 0,
    turnSkipped: 0,
    turnEvents: [],
    lastTurnEndReason: null,
    history: [],
    score: scoring.create(teamIds),
    endRequested: false,
    winnerTeamIds: [],
    errorCode,
  };
}

export function reducer(state: PantomimeState, action: PantomimeAction): PantomimeState {
  const s = state;
  switch (action.type) {
    case 'HANDOFF_READY': {
      if (s.phase !== 'handoff') return s;
      const base: PantomimeState = {
        ...s,
        turnCorrect: 0,
        turnSkipped: 0,
        turnEvents: [],
        lastTurnEndReason: null,
        gate: revealGate.init([asPlayerId(actorId(s))]),
        clock: timer.create('countdown', s.options.roundSeconds * 1000),
      };
      const { deck: d, promptId } = drawNext(base.deck);
      if (promptId === null) {
        return finalizeTurn({ ...base, currentPromptId: null }, 'deckExhausted', 0);
      }
      return {
        ...base,
        deck: d,
        currentPromptId: promptId,
        phase: phaseMachine.go(MACHINE, s.phase, 'reveal'),
      };
    }
    case 'REVEAL': {
      if (s.phase !== 'reveal') return s;
      return { ...s, gate: revealGate.reveal(s.gate) };
    }
    case 'START_ACTING': {
      if (s.phase !== 'reveal') return s;
      return {
        ...s,
        phase: phaseMachine.go(MACHINE, s.phase, 'acting'),
        clock: timer.start(
          timer.create('countdown', s.options.roundSeconds * 1000),
          action.now,
        ),
      };
    }
    case 'TICK': {
      if (s.phase !== 'acting') return s;
      const clock = timer.tick(s.clock, action.now);
      if (timer.isExpired(clock, action.now)) {
        return finalizeTurn({ ...s, clock }, 'timeExpired', action.now);
      }
      return clock === s.clock ? s : { ...s, clock };
    }
    case 'CORRECT': {
      if (s.phase !== 'acting' || s.currentPromptId === null) return s;
      const teamIdx = s.turn.index;
      const teamId = s.teams[teamIdx].teamId;
      const turnEvents: TurnEvent[] = [
        ...s.turnEvents,
        { promptId: s.currentPromptId, result: 'correct' },
      ];
      const teams = updateTeam(s.teams, teamIdx, (t) => ({
        ...t,
        correctCount: t.correctCount + 1,
      }));
      const score = scoring.add(s.score, teamId, 1, 'correct', 0);
      const endRequested =
        s.endRequested ||
        (s.options.endMode === 'targetScore' &&
          scoring.total(score, teamId) >= s.options.targetScore);
      const discarded = deck.discard(s.deck, s.currentPromptId);
      const { deck: d, promptId } = drawNext(discarded);
      const next: PantomimeState = {
        ...s,
        teams,
        score,
        endRequested,
        turnCorrect: s.turnCorrect + 1,
        turnEvents,
      };
      if (promptId === null) {
        return finalizeTurn({ ...next, deck: d, currentPromptId: null }, 'deckExhausted', 0);
      }
      return { ...next, deck: d, currentPromptId: promptId };
    }
    case 'SKIP': {
      if (s.phase !== 'acting' || s.currentPromptId === null) return s;
      // Skip-cap guard: no-op once the cap is reached (unless unlimited).
      if (s.options.maxSkipsPerTurn !== -1 && s.turnSkipped >= s.options.maxSkipsPerTurn) {
        return s;
      }
      const teamIdx = s.turn.index;
      const teamId = s.teams[teamIdx].teamId;
      const turnEvents: TurnEvent[] = [
        ...s.turnEvents,
        { promptId: s.currentPromptId, result: 'skip' },
      ];
      const teams = updateTeam(s.teams, teamIdx, (t) => ({ ...t, skipCount: t.skipCount + 1 }));
      let score = s.score;
      if (s.options.skipPenalty && scoring.total(score, teamId) > 0) {
        score = scoring.add(score, teamId, -1, 'skipPenalty', 0);
      }
      const discarded = deck.discard(s.deck, s.currentPromptId);
      const { deck: d, promptId } = drawNext(discarded);
      const next: PantomimeState = {
        ...s,
        teams,
        score,
        turnSkipped: s.turnSkipped + 1,
        turnEvents,
      };
      if (promptId === null) {
        return finalizeTurn({ ...next, deck: d, currentPromptId: null }, 'deckExhausted', 0);
      }
      return { ...next, deck: d, currentPromptId: promptId };
    }
    case 'PAUSE': {
      if (s.phase !== 'acting') return s;
      return { ...s, clock: timer.pause(s.clock, action.now) };
    }
    case 'RESUME': {
      if (s.phase !== 'acting') return s;
      return { ...s, clock: timer.start(s.clock, action.now) };
    }
    case 'END_TURN_EARLY': {
      if (s.phase !== 'acting') return s;
      return finalizeTurn(s, 'manualEnd', action.now);
    }
    case 'NEXT_TURN': {
      if (s.phase !== 'turnEnd') return s;
      const playedIdx = s.turn.index;
      const teams = updateTeam(s.teams, playedIdx, (t) => ({
        ...t,
        actorCursor: t.playerIds.length ? (t.actorCursor + 1) % t.playerIds.length : 0,
      }));
      const turn = turnOrder.next(s.turn, action.seed);
      const roundComplete = turn.index === 0; // wrapped back to the first team
      const finish =
        s.options.endMode === 'rounds'
          ? turn.round >= s.options.totalRounds
          : s.endRequested && roundComplete;
      if (finish) {
        const res = results.fromScores(s.score);
        return {
          ...s,
          teams,
          turn,
          phase: 'results',
          finished: true,
          winnerTeamIds: res.winners,
        };
      }
      return {
        ...s,
        teams,
        turn,
        phase: phaseMachine.go(MACHINE, s.phase, 'handoff'),
        currentPromptId: null,
        turnCorrect: 0,
        turnSkipped: 0,
        turnEvents: [],
        lastTurnEndReason: null,
        gate: revealGate.init([]),
        clock: timer.create('countdown', s.options.roundSeconds * 1000),
      };
    }
    case 'RESET':
      return s; // no-op; "play again" is host-driven (re-creates with a fresh seed)
    default:
      return s;
  }
}

/* ─────────────────────────  Pure selectors  ───────────────────────── */

export const currentRound = (s: PantomimeState): number => s.turn.round + 1;
export const skipsLeft = (s: PantomimeState): number =>
  s.options.maxSkipsPerTurn === -1 ? Infinity : s.options.maxSkipsPerTurn - s.turnSkipped;
export const selectStandings = (s: PantomimeState) => results.fromScores(s.score).standings;
export const selectWinners = (s: PantomimeState): string[] => results.fromScores(s.score).winners;
export const teamLabel = (s: PantomimeState, teamId: string): string =>
  s.teams.find((t) => t.teamId === teamId)?.name ?? teamId;
export const teamColor = (s: PantomimeState, teamId: string): ColorToken | undefined =>
  s.teams.find((t) => t.teamId === teamId)?.color;

File: android-native/app/src/main/java/com/gamenight/party/game/pantomime/Logic.kt
package com.gamenight.party.game.pantomime

import com.gamenight.party.engine.DeckState
import com.gamenight.party.engine.PhaseMachine
import com.gamenight.party.engine.PhaseNode
import com.gamenight.party.engine.RevealGateState
import com.gamenight.party.engine.ScoreState
import com.gamenight.party.engine.Standing
import com.gamenight.party.engine.TimerMode
import com.gamenight.party.engine.TimerState
import com.gamenight.party.engine.TurnMode
import com.gamenight.party.engine.TurnOrderState
import com.gamenight.party.engine.add
import com.gamenight.party.engine.asPlayerId
import com.gamenight.party.engine.create
import com.gamenight.party.engine.defineMachine
import com.gamenight.party.engine.discard
import com.gamenight.party.engine.draw
import com.gamenight.party.engine.fromScores
import com.gamenight.party.engine.go
import com.gamenight.party.engine.init
import com.gamenight.party.engine.isExpired
import com.gamenight.party.engine.next
import com.gamenight.party.engine.pause
import com.gamenight.party.engine.reveal
import com.gamenight.party.engine.start
import com.gamenight.party.engine.tick
import com.gamenight.party.engine.total
import com.gamenight.party.model.ColorToken

/**
 * PURE game logic for Pantomime — a faithful Kotlin port of src/games/pantomime/logic.ts. No
 * clock / RNG / IO: timestamps (`now`) and shuffle seeds arrive in actions / [createInitialState].
 * Side effects (sound/haptics/the wall clock) live in the Compose layer, never here. No-op cases
 * return the SAME state instance to mirror the web reducer's `=== input` semantics.
 */

enum class PantomimePhase {
    HANDOFF, // "Pass the phone to <Actor> of <Team>"
    REVEAL,  // gate closed: actor taps to privately read the prompt
    ACTING,  // prompt revealed, timer running, team guessing
    TURN_END, // turn over: show turn summary
    RESULTS, // game over
    ERROR,   // bad config / empty pool (never reachable from a valid Setup)
}

enum class TurnEndReason { TIME_EXPIRED, DECK_EXHAUSTED, MANUAL_END }

enum class TurnResult { CORRECT, SKIP }

enum class PantomimeErrorCode { EMPTY_DECK, BAD_TEAMS }

data class TurnEvent(val promptId: String, val result: TurnResult)

data class PantomimeTeamState(
    val teamId: String,
    val name: String,
    val color: ColorToken? = null,
    val playerIds: List<String>,
    /** Index into [playerIds] of who acts on this team's NEXT turn. */
    val actorCursor: Int,
    val correctCount: Int,
    val skipCount: Int,
)

data class PantomimeTurnRecord(
    val turnIndex: Int, // global, 0-based
    val roundIndex: Int, // 0-based
    val teamId: String,
    val actorId: String,
    val correct: Int,
    val skipped: Int,
    val endReason: TurnEndReason,
    val promptIds: List<String>, // resolved this turn (correct + skipped)
)

data class PantomimeState(
    val v: Int = 1,
    val phase: PantomimePhase,
    val finished: Boolean,
    val options: PantomimeOptions,
    val playerNames: Map<String, String>,
    /** Teams in fixed turn order. */
    val teams: List<PantomimeTeamState>,
    /** Turn order over teamIds (index = active team, round = completed cycles). */
    val turn: TurnOrderState,
    val deck: DeckState<String>,
    val currentPromptId: String?,
    val gate: RevealGateState,
    val clock: TimerState,
    val turnCorrect: Int,
    val turnSkipped: Int,
    val turnEvents: List<TurnEvent>,
    val lastTurnEndReason: TurnEndReason?,
    val history: List<PantomimeTurnRecord>,
    val score: ScoreState, // keyed by teamId
    /** Set when a target-score end condition fires; the round still finishes (fairness). */
    val endRequested: Boolean,
    val winnerTeamIds: List<String>,
    val errorCode: PantomimeErrorCode?,
)

sealed interface PantomimeAction {
    data object HandoffReady : PantomimeAction
    data object Reveal : PantomimeAction
    data class StartActing(val now: Long) : PantomimeAction
    data class Tick(val now: Long) : PantomimeAction
    data object Correct : PantomimeAction
    data object Skip : PantomimeAction
    data class Pause(val now: Long) : PantomimeAction
    data class Resume(val now: Long) : PantomimeAction
    data class EndTurnEarly(val now: Long) : PantomimeAction
    data class NextTurn(val seed: Int) : PantomimeAction
    data object Reset : PantomimeAction
}

private val MACHINE: PhaseMachine<PantomimePhase> = defineMachine(
    PhaseMachine(
        initial = PantomimePhase.HANDOFF,
        nodes = mapOf(
            PantomimePhase.HANDOFF to PhaseNode(PantomimePhase.HANDOFF, listOf(PantomimePhase.REVEAL, PantomimePhase.TURN_END)),
            PantomimePhase.REVEAL to PhaseNode(PantomimePhase.REVEAL, listOf(PantomimePhase.ACTING, PantomimePhase.TURN_END)),
            PantomimePhase.ACTING to PhaseNode(PantomimePhase.ACTING, listOf(PantomimePhase.TURN_END)),
            PantomimePhase.TURN_END to PhaseNode(PantomimePhase.TURN_END, listOf(PantomimePhase.HANDOFF, PantomimePhase.RESULTS)),
            PantomimePhase.RESULTS to PhaseNode(PantomimePhase.RESULTS, emptyList(), terminal = true),
            PantomimePhase.ERROR to PhaseNode(PantomimePhase.ERROR, emptyList(), terminal = true),
        ),
    ),
)

private val TEAM_COLORS: List<ColorToken> =
    listOf(ColorToken.ROSE, ColorToken.SKY, ColorToken.LIME, ColorToken.GOLD)

/** Seed used for in-game deck reshuffles. Deterministic (the initial shuffle uses the host seed). */
private const val DRAW_SEED: Int = 0

private fun drawNext(d: DeckState<String>): Pair<DeckState<String>, String?> {
    val r = draw(d, 1, DRAW_SEED)
    return r.deck to r.drawn.firstOrNull()
}

private fun updateTeam(
    teams: List<PantomimeTeamState>,
    index: Int,
    patch: (PantomimeTeamState) -> PantomimeTeamState,
): List<PantomimeTeamState> = teams.mapIndexed { i, t -> if (i == index) patch(t) else t }

// ──────────────────────────── Pure selectors ────────────────────────────

fun activeTeam(s: PantomimeState): PantomimeTeamState? = s.teams.getOrNull(s.turn.index)

fun actorId(s: PantomimeState): String {
    val team = s.teams.getOrNull(s.turn.index) ?: return ""
    return team.playerIds.getOrNull(team.actorCursor) ?: ""
}

fun actorName(s: PantomimeState): String = s.playerNames[actorId(s)] ?: ""

fun currentRound(s: PantomimeState): Int = s.turn.round + 1

/** Skips left this turn; [Int.MAX_VALUE] models the web's `Infinity` for unlimited skips. */
fun skipsLeft(s: PantomimeState): Int =
    if (s.options.maxSkipsPerTurn == -1) Int.MAX_VALUE else s.options.maxSkipsPerTurn - s.turnSkipped

fun selectStandings(s: PantomimeState): List<Standing> = fromScores(s.score).standings

fun selectWinners(s: PantomimeState): List<String> = fromScores(s.score).winners

fun teamLabel(s: PantomimeState, teamId: String): String =
    s.teams.firstOrNull { it.teamId == teamId }?.name ?: teamId

fun teamColor(s: PantomimeState, teamId: String): ColorToken? =
    s.teams.firstOrNull { it.teamId == teamId }?.color

// ──────────────────────────── Finalize / init ────────────────────────────

private fun finalizeTurn(s: PantomimeState, reason: TurnEndReason, now: Long): PantomimeState {
    val team = s.teams[s.turn.index]
    // Return any unresolved current prompt to the draw pile (not counted, not lost).
    var d = s.deck
    val current = s.currentPromptId
    if (current != null) {
        d = s.deck.copy(drawPile = s.deck.drawPile + current)
    }
    val record = PantomimeTurnRecord(
        turnIndex = s.history.size,
        roundIndex = s.turn.round,
        teamId = team.teamId,
        actorId = actorId(s),
        correct = s.turnCorrect,
        skipped = s.turnSkipped,
        endReason = reason,
        promptIds = s.turnEvents.map { it.promptId },
    )
    return s.copy(
        phase = PantomimePhase.TURN_END,
        deck = d,
        currentPromptId = null,
        lastTurnEndReason = reason,
        history = s.history + record,
        clock = pause(s.clock, now),
    )
}

fun createInitialState(config: PantomimeConfig, content: PantomimeContent, seed: Int): PantomimeState {
    val options = normalizeOptions(config.options)
    val playerNames: Map<String, String> = config.players.associate { it.id to it.name }

    val teams: List<PantomimeTeamState> = config.teams.mapIndexed { i, t ->
        PantomimeTeamState(
            teamId = t.id,
            name = t.name.ifBlank { "Team ${i + 1}" },
            color = TEAM_COLORS[i % TEAM_COLORS.size],
            playerIds = t.memberIds,
            actorCursor = 0,
            correctCount = 0,
            skipCount = 0,
        )
    }

    val pool = buildPool(content, options)
    val deckState = create(pool.map { it.id }, seed)
    val teamIds = teams.map { it.teamId }
    val turn = init(teamIds, TurnMode.CIRCULAR, seed)

    val badTeams = teams.size < 2 || teams.any { it.playerIds.size < 2 }
    val empty = pool.isEmpty()
    val errorCode: PantomimeErrorCode? = when {
        badTeams -> PantomimeErrorCode.BAD_TEAMS
        empty -> PantomimeErrorCode.EMPTY_DECK
        else -> null
    }

    return PantomimeState(
        v = 1,
        phase = if (errorCode != null) PantomimePhase.ERROR else PantomimePhase.HANDOFF,
        finished = false,
        options = options,
        playerNames = playerNames,
        teams = teams,
        turn = turn,
        deck = deckState,
        currentPromptId = null,
        gate = init(emptyList<String>()),
        clock = create(TimerMode.COUNTDOWN, options.roundSeconds * 1000L),
        turnCorrect = 0,
        turnSkipped = 0,
        turnEvents = emptyList(),
        lastTurnEndReason = null,
        history = emptyList(),
        score = create(teamIds),
        endRequested = false,
        winnerTeamIds = emptyList(),
        errorCode = errorCode,
    )
}

// ──────────────────────────── Reducer ────────────────────────────

fun reducer(state: PantomimeState, action: PantomimeAction): PantomimeState {
    val s = state
    return when (action) {
        is PantomimeAction.HandoffReady -> {
            if (s.phase != PantomimePhase.HANDOFF) return s
            val base = s.copy(
                turnCorrect = 0,
                turnSkipped = 0,
                turnEvents = emptyList(),
                lastTurnEndReason = null,
                gate = init(listOf(asPlayerId(actorId(s)))),
                clock = create(TimerMode.COUNTDOWN, s.options.roundSeconds * 1000L),
            )
            val (d, promptId) = drawNext(base.deck)
            if (promptId == null) {
                finalizeTurn(base.copy(currentPromptId = null), TurnEndReason.DECK_EXHAUSTED, 0L)
            } else {
                base.copy(
                    deck = d,
                    currentPromptId = promptId,
                    phase = go(MACHINE, s.phase, PantomimePhase.REVEAL),
                )
            }
        }

        is PantomimeAction.Reveal -> {
            if (s.phase != PantomimePhase.REVEAL) return s
            s.copy(gate = reveal(s.gate))
        }

        is PantomimeAction.StartActing -> {
            if (s.phase != PantomimePhase.REVEAL) return s
            s.copy(
                phase = go(MACHINE, s.phase, PantomimePhase.ACTING),
                clock = start(create(TimerMode.COUNTDOWN, s.options.roundSeconds * 1000L), action.now),
            )
        }

        is PantomimeAction.Tick -> {
            if (s.phase != PantomimePhase.ACTING) return s
            val clock = tick(s.clock, action.now)
            if (isExpired(clock, action.now)) {
                finalizeTurn(s.copy(clock = clock), TurnEndReason.TIME_EXPIRED, action.now)
            } else if (clock === s.clock) {
                s
            } else {
                s.copy(clock = clock)
            }
        }

        is PantomimeAction.Correct -> {
            val current = s.currentPromptId
            if (s.phase != PantomimePhase.ACTING || current == null) return s
            val teamIdx = s.turn.index
            val teamId = s.teams[teamIdx].teamId
            val turnEvents = s.turnEvents + TurnEvent(current, TurnResult.CORRECT)
            val teams = updateTeam(s.teams, teamIdx) { it.copy(correctCount = it.correctCount + 1) }
            val score = add(s.score, teamId, 1, "correct", 0L)
            val endRequested = s.endRequested ||
                (s.options.endMode == PantomimeEndMode.TARGET_SCORE && total(score, teamId) >= s.options.targetScore)
            val discarded = discard(s.deck, current)
            val (d, promptId) = drawNext(discarded)
            val next = s.copy(
                teams = teams,
                score = score,
                endRequested = endRequested,
                turnCorrect = s.turnCorrect + 1,
                turnEvents = turnEvents,
            )
            if (promptId == null) {
                finalizeTurn(next.copy(deck = d, currentPromptId = null), TurnEndReason.DECK_EXHAUSTED, 0L)
            } else {
                next.copy(deck = d, currentPromptId = promptId)
            }
        }

        is PantomimeAction.Skip -> {
            val current = s.currentPromptId
            if (s.phase != PantomimePhase.ACTING || current == null) return s
            // Skip-cap guard: no-op once the cap is reached (unless unlimited).
            if (s.options.maxSkipsPerTurn != -1 && s.turnSkipped >= s.options.maxSkipsPerTurn) return s
            val teamIdx = s.turn.index
            val teamId = s.teams[teamIdx].teamId
            val turnEvents = s.turnEvents + TurnEvent(current, TurnResult.SKIP)
            val teams = updateTeam(s.teams, teamIdx) { it.copy(skipCount = it.skipCount + 1) }
            var score = s.score
            if (s.options.skipPenalty && total(score, teamId) > 0) {
                score = add(score, teamId, -1, "skipPenalty", 0L)
            }
            val discarded = discard(s.deck, current)
            val (d, promptId) = drawNext(discarded)
            val next = s.copy(
                teams = teams,
                score = score,
                turnSkipped = s.turnSkipped + 1,
                turnEvents = turnEvents,
            )
            if (promptId == null) {
                finalizeTurn(next.copy(deck = d, currentPromptId = null), TurnEndReason.DECK_EXHAUSTED, 0L)
            } else {
                next.copy(deck = d, currentPromptId = promptId)
            }
        }

        is PantomimeAction.Pause -> {
            if (s.phase != PantomimePhase.ACTING) return s
            s.copy(clock = pause(s.clock, action.now))
        }

        is PantomimeAction.Resume -> {
            if (s.phase != PantomimePhase.ACTING) return s
            s.copy(clock = start(s.clock, action.now))
        }

        is PantomimeAction.EndTurnEarly -> {
            if (s.phase != PantomimePhase.ACTING) return s
            finalizeTurn(s, TurnEndReason.MANUAL_END, action.now)
        }

        is PantomimeAction.NextTurn -> {
            if (s.phase != PantomimePhase.TURN_END) return s
            val playedIdx = s.turn.index
            val teams = updateTeam(s.teams, playedIdx) {
                it.copy(actorCursor = if (it.playerIds.isNotEmpty()) (it.actorCursor + 1) % it.playerIds.size else 0)
            }
            val turn = next(s.turn, action.seed)
            val roundComplete = turn.index == 0 // wrapped back to the first team
            val finish = if (s.options.endMode == PantomimeEndMode.ROUNDS) {
                turn.round >= s.options.totalRounds
            } else {
                s.endRequested && roundComplete
            }
            if (finish) {
                val res = fromScores(s.score)
                s.copy(
                    teams = teams,
                    turn = turn,
                    phase = PantomimePhase.RESULTS,
                    finished = true,
                    winnerTeamIds = res.winners,
                )
            } else {
                s.copy(
                    teams = teams,
                    turn = turn,
                    phase = go(MACHINE, s.phase, PantomimePhase.HANDOFF),
                    currentPromptId = null,
                    turnCorrect = 0,
                    turnSkipped = 0,
                    turnEvents = emptyList(),
                    lastTurnEndReason = null,
                    gate = init(emptyList<String>()),
                    clock = create(TimerMode.COUNTDOWN, s.options.roundSeconds * 1000L),
                )
            }
        }

        is PantomimeAction.Reset -> s // no-op; "play again" is host-driven (re-create with a fresh seed)
    }
}
